packages/ts/generator-typescript-plugin-model/src/ModelSchemaProcessor.ts

```diff
@@ -0,0 +1,200 @@
+    } else if (isArraySchema(unwrappedSchema)) {
+      [type, modelType, model, args] = this.#processArray(unwrappedSchema);
+    } else if (isMapSchema(unwrappedSchema)) {
+      [type, modelType, model, args] = this.#processRecord(unwrappedSchema);
+    } else if (isStringSchema(unwrappedSchema)) {
+      type = ts.factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword);
+      model = this.#getBuiltinFormExport('StringModel');
+      modelType = ts.factory.createTypeReferenceNode(model);
+    } else if (isNumberSchema(unwrappedSchema) || isIntegerSchema(unwrappedSchema)) {
+      type = ts.factory.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword);
+      model = this.#getBuiltinFormExport('NumberModel');
+      modelType = ts.factory.createTypeReferenceNode(model);
+    } else if (isBooleanSchema(unwrappedSchema)) {
+      type = ts.factory.createKeywordTypeNode(ts.SyntaxKind.BooleanKeyword);
+      model = this.#getBuiltinFormExport('BooleanModel');
+      modelType = ts.factory.createTypeReferenceNode(model);
+    } else {
+      type = ts.factory.createKeywordTypeNode(ts.SyntaxKind.UnknownKeyword);
+      model = this.#getBuiltinFormExport('ObjectModel');
+      modelType = ts.factory.createTypeReferenceNode(model);
+    }
+
+    const optionalArg = isNullableSchema(this.#schema) ? ts.factory.createTrue() : ts.factory.createFalse();
+
+    if (isAnnotatedSchema(unwrappedSchema)) {
+      args = [...args, ...this.#getValidators(unwrappedSchema)];
+    }
+
+    return [type, modelType, model, ts.factory.createArrayLiteralExpression([optionalArg, ...args])];
+  }
+
+  #processReference(schema: ReferenceSchema): [TypeNode, TypeNode, Identifier, Expression[]] {
+    const schemaPath = convertReferenceSchemaToPath(schema);
+    const typeName = convertReferenceSchemaToSpecifier(schema);
+    const typePath = this.#dependencies.paths.createRelativePath(schemaPath, this.#cwd);
+    const modelPath = this.#dependencies.paths.createRelativePath(`${schemaPath}Model`, this.#cwd);
+    const modelName = `${typeName}Model`;
+    const refType =
+      this.#dependencies.imports.default.getIdentifier(typePath) ??
+      this.#dependencies.imports.default.add(typePath, typeName, true);
+    const type = ts.factory.createTypeReferenceNode(refType);
+    const model =
+      this.#dependencies.imports.default.getIdentifier(modelPath) ??
+      this.#dependencies.imports.default.add(modelPath, modelName);
+    const modelType = ts.factory.createTypeReferenceNode(model);
+    return [type, modelType, model, []];
+  }
+
+  #processArray(schema: ArraySchema): [TypeNode, TypeNode, Identifier, Expression[]] {
+    const model = this.#getBuiltinFormExport('ArrayModel');
+    const [itemType, itemModelType, itemModel, itemArgs] = new ModelSchemaProcessor(
+      schema.items,
+      this.#dependencies,
+      this.#cwd,
+    ).process();
+    const type = ts.factory.createTypeReferenceNode(ts.factory.createIdentifier('ReadonlyArray'), [itemType]);
```

Comment:
Correct me if I'm wrong but shouldn't we use a regular `Array` here? I thought all the `ReadonlyArray` structures were removed with readonly change rollback.

Reply:
Regardless of that readonly change and rollback, in form models we used `ReadonlyArray` from the start. The preferred way of mutating arrays in the Binder is using Binder APIs.